---
title: "3'UTR length distribution among eukaryotes"
output: html_document
date: "2024"
---

```{r}
library(biomaRt)
library(AnnotationHub)
library(orthogene)
library(tidyverse)
library(ggridges)
library(ggpubr)
library(rstatix)
library(hrbrthemes)
library(dplyr)
library(readr)
library(DESeq2)
library(clusterProfiler)
library(GOfuncR)
library(GO.db)
```

Visualization functions

```{r}
boxoploto <- function(data_for_plot, zoom=1){
  data_for_plot %>% ggplot(aes(x = Codon, y = Length, color = Codon)) +
    geom_boxplot(lwd=1.1, fatten = 1)+
    coord_cartesian(ylim = quantile(data_for_plot$Length, c(0, zoom)))+
    scale_color_manual(values = c("#ff7f00","#007fff","#ff0065"))+
    theme_minimal()+
    theme(text = element_text(size = 25), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)), axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l=0)), axis.line.x = element_line(colour = 'black', linewidth = 0.5, linetype='solid'),
          axis.line.y = element_line(colour = 'black', linewidth = 0.5, linetype='solid'), axis.ticks = element_line(colour = 'black'), axis.ticks.length=unit(0.2,"cm"))
} #function for plotting alike boxplots with zoom option, enabling scaling. zoom takes values from 0 to 1 representing the part of the data to be plotted (from 0 to zoom)

densoploto <- function(data_for_plot,maxlength = 1000000000){
  data_for_plot %>% filter (Length < maxlength) %>% ggplot(aes(x=Length, color=Codon))+
  geom_density(adjust=1.5, size=0.8)+
  scale_color_manual(values = c("#ff7f00","#007fff","#ff0065"))+
  theme_ipsum(base_family = "Arial", base_size = 20, axis_title_size = 25,axis=TRUE, ticks=TRUE)+
  theme(text = element_text(size = 25), axis.text=element_text(size=25), axis.title = element_text(size=25),panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line.x = element_line(colour = 'black', linewidth = 0.5, linetype='solid'),
          axis.line.y = element_line(colour = 'black', linewidth = 0.5, linetype='solid'), axis.ticks = element_line(colour = 'black'), axis.ticks.length=unit(0.2,"cm"))+
    xlab("Length")+ ylab("Density")
} #function for plotting alike probability density plots with maxlength option, enabling to set the limit of 3'UTR length to be included to the plot
```


##Human data analysis

###General distribution

```{r}
human_data <- read_delim("../data/RES_Homo_sapiens.GRCh38.107.utrs.tsv",delim="\t", col_names = c("Transcript_ID", "Codon", "Length"))
human_data$Codon <- gsub("T", "U", human_data$Codon)

human_data%>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
human_data %>% group_by(Codon) %>% dplyr::count()
```

```{r}
human_box <- boxoploto(human_data) 
human_box_09 <- boxoploto(human_data, 0.9)
human_dens <- densoploto(human_data)
human_dens_zoom <- densoploto(human_data, 3000)
```

###Human enrichment analysis

####Small 3'UTRs

Two groups: L<=10 and 10<L<=50

```{r}
hum_total <- read_delim("../data/Human_total_list.tsv", delim = "\t", col_names = c("Transcript_ID", "Gene"))
hum_trans_gene <- merge(human_data, hum_total, by="Transcript_ID") %>% dplyr::select(Gene,Codon,Length)
hum_10 <- hum_trans_gene %>% filter(Length<=10)

hum_10_go <- enrichGO(gene = hum_10$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "ALL",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)


hum_11_50 <- hum_trans_gene %>% filter(Length>10 & Length <=50)
hum_11_50_go <- enrichGO(gene = hum_11_50$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "ALL",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)


hum_10_go_emap <- hum_10_go %>% enrichplot::pairwise_termsim() %>%
  emapplot(showCategory = 50,cex_label_category = 0.8, cex_category = 0.5, repel = TRUE)

hum_11_50_go_emap <- hum_11_50_go %>% enrichplot::pairwise_termsim() %>%
emapplot(showCategory = 50,cex_label_category = 1, cex_category = 0.5, repel=TRUE, label_format  = 0.5) + theme(text = element_text(size = 25), legend.text=element_text(size=14), legend.title=element_text(size=14)) + scale_fill_gradientn(colours=c("#ff7f00", "#007fff"), guide = "colorbar", aesthetics = c("fill")) + labs(fill = "p-adjust")

```


And now let's obtain genes that are located in the nodes of interest (ribosomes, translation) and have only short 3'UTRs. Since single gene can have several transcript variants, it can have 3'UTRs both <50 and >100. Here we are interested in the genes having only small variants. Those that have both short and long 3'UTRs are basically available from the results of enrichment analysis.  

First was to filter the letter files to obtain genes with only interesting GO categories and than get the list of unique genes, as they can occur in several GO categories. This was made with Python script ../scripts_analysis/gene_extractor.py

```{r}
hum_11_50_genes <- read_delim ("../results/GENES_50_HUM.tsv", delim = "\t", col_names = c("Gene"))#there are also genes that also have long 3'UTRs, let's get rid of them

hum_11_50_genes_ensembl <- getBM(c("external_gene_name","ensembl_gene_id"),"external_gene_name", hum_11_50_genes$Gene, mart)

colnames(hum_11_50_genes_ensembl) <-  c("Symbol", "Gene")

hum_11_50_trans_length_ensembl <- merge(hum_11_50_genes_ensembl, hum_trans_gene, by = "Gene")

less <-  hum_11_50_trans_length_ensembl %>% filter (Length <= 50) 
more <-  hum_11_50_trans_length_ensembl %>% filter (Length > 50) 
small <- less %>% filter(!Symbol %in% more$Symbol) %>% dplyr::select("Gene", "Symbol", "Length")
```



###Housekeeping genes

List of human housekeeping genes is taken from https://www.tau.ac.il/~elieis/HKG/. It is stored in the "../data/HK_genes.txt" file

```{r}
hum_house <- read_delim("../data/HK_genes.txt", delim = " \t", col_names = c("SYMBOL", "Nucl_DB"))

hum_house_ensembl <- getBM("ensembl_gene_id","external_gene_name", hum_house$SYMBOL, mart)
colnames(hum_house_ensembl) <- c("Gene")
hum_house_merged <- merge(hum_house_ensembl, hum_trans_gene, by="Gene")

hum_house_merged$Codon <- gsub("T", "U", hum_house_merged$Codon)

hum_house_box <- boxoploto(hum_house_merged)
hum_house_box_09 <- boxoploto(hum_house_merged, 0.9)
hum_house_dens <- densoploto(hum_house_merged)
hum_house_dens_zoom <- densoploto(hum_house_merged, 3000)

hum_house_merged %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
```

##Rabbit data analysis

###Rabbit vs Human


Primarily, it's required to obtain the list of rabbit and human orthologous genes. 

```{r}
mart <- useMart("ENSEMBL_MART_ENSEMBL","hsapiens_gene_ensembl") #human mart object
```

Get ENSEMBL gene IDs from transcript IDs
```{r}
rab_res <- read_delim("../data/RES_Oryctolagus_cuniculus.OryCun2.0.107.utrs.tsv",delim="\t",col_names = c("Transcript_ID","Codon","Length"))
rab_total <- read_delim("../data/Oryctolagus.tsv",delim="\t",col_names = c("Symbol","Gene","Transcript_ID"))

merged_res <- merge(rab_res,rab_total,by="Transcript_ID")
rab_genes <- merged_res %>% dplyr::select(Gene)
colnames(rab_genes) <- c("Rab_Gene")

```

```{r}
rab_mart <- useMart("ensembl", dataset = "ocuniculus_gene_ensembl", host="https://feb2021.archive.ensembl.org")
human_mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl",host = "https://feb2021.archive.ensembl.org")

orth <- getLDS(mart = rab_mart, attributes = c("ensembl_gene_id"), martL = human_mart, attributesL = c("ensembl_gene_id"))
colnames(orth) <- c("Rab_Gene", "Hum_Gene")

```

```{r}
target_orths <- merge(rab_genes, orth, by = "Rab_Gene") #have some replicated rows, so has to be filtered
```


```{r}
uniq_orths <- target_orths %>% unique()
target_genes_rabbit <- uniq_orths %>% dplyr::select(Rab_Gene)
target_genes_human <- uniq_orths %>% dplyr::select(Hum_Gene)
colnames(target_genes_rabbit) <- c("Gene")
colnames(target_genes_human) <- c("Gene")
```

Now we have a list of orthologous genes. Next step is to get the list of transcripts for each organism. To do that tables provided by the UTRdb 2.0 will be used

```{r}
target_rabbit_transcripts <- merge(target_genes_rabbit, rab_total, by = "Gene") %>% dplyr::select("Transcript_ID") %>% unique()
target_human_transcripts <- merge(target_genes_human, hum_total, by = "Gene") %>% dplyr::select("Transcript_ID") %>% unique()
```

These variables (target_*_transcripts) were written to tsv files, filtered from replicated values final lists of transcripts were used to obtain information about 3'UTR lengths. This was made with Python script ../scripts_data/UTR_orth.py

Results are saved in the files RES_Human and RES_Rabbit. 
```{r}
hum_orth <- read_delim("../data/ORTHRES_HT.tsv",delim="\t",col_names = c("Trans_ID", "Codon","Length"))
rab_orth <- read_delim("../data/ORTHRES_RT.tsv",delim="\t",col_names = c("Trans_ID", "Codon","Length"))

hum_orth$Codon <- gsub("T", "U",hum_orth$Codon )
rab_orth$Codon <- gsub("T", "U",rab_orth$Codon )

h_o_box <- boxoploto(hum_orth) 
r_o_box <- boxoploto(rab_orth)
h_o_box_zoomed <- boxoploto(hum_orth,0.9)
r_o_box_zoomed <- boxoploto(rab_orth,0.9)
h_o_dens <- densoploto(hum_orth,3000)
r_o_dens <- densoploto(rab_orth)

hum_orth %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
rab_orth %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))

hum_orth %>% group_by(Codon) %>% dplyr::count()
rab_orth %>% group_by(Codon) %>% dplyr::count()
```


##General distribution

```{r}
total_prec <- read_delim(unz("../data/TOTAL.tsv.zip","TOTAL.tsv"),delim = "\t", col_names = c("Transcript_ID", "Codon", "Length")) #TOTAL_PY.tsv is too big, so it's kept as a .zip file in the repo
total_prec$Codon <- gsub("T","U", total_prec$Codon)


prec_dens <- densoploto(total_prec)
prec_box <- boxoploto(total_prec)
prec_box_zoom <- boxoploto(total_prec, 0.9)
prec_dens_zoom <- densoploto(total_prec, 2000)

prec_plot %>% nrow()
prec_plot %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
prec_plot %>% group_by(Codon) %>% dplyr::count()
```

##Small 3'UTR holders

###Paramecium tetraurelia

Data is obtained with Python script ../scripts_data/Paramecium.py

```{r}
param_data <- read_delim("../data/Paramecium_tetraurelia.RES.tsv", delim="\t", col_names = c("Transcript_ID", "Codon", "Length"))

param_data$Codon <- gsub("T", "U", param_data$Codon)

param_box <- boxoploto(param_data %>% filter (Codon =="UGA"))
param_box_nonzero <- boxoploto(param_data %>% filter (Length> 0))
param_box_nonzero_95 <- boxoploto(param_data %>% filter (Length> 0), 0.95)
param_dens <- densoploto(param_data %>% filter (Codon =="UGA"))

param_data %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
param_data %>% group_by(Codon) %>% dplyr::count()

param_uga <- param_data %>% filter(Codon == "UGA")
```

###Others
Organisms having small 3'UTR length median values were chosen with Python script ../scripts_analysis/small_utrs.py

```{r}

codons_T <- c("TAA", "TAG", "TGA")
o_lucim_data <- read_delim("../data/RES_Ostreococcus_lucimarinus.ASM9206v1.54.utrs.tsv", delim = "\t",col_names = c("Transcript_ID", "Codon","Length"))
m_lucif_data <- read_delim("../data/RES_Myotis_lucifugus.Myoluc2.0.107.utrs.tsv", delim = "\t",col_names = c("Transcript_ID", "Codon","Length"))
a_lyrata_data <- read_delim("../data/RES_Arabidopsis_lyrata.v.1.0.54.utrs.tsv", delim = "\t", col_names = c("Transcript_ID", "Codon","Length"))
d_willistoni_data <- read_delim("../data/RES_Drosophila_willistoni.dwil_caf1.54.utrs.tsv", delim = "\t", col_names = c("Transcript_ID", "Codon","Length")) 

o_lucim_data$Codon <- gsub("T", "U", o_lucim_data$Codon)
m_lucif_data$Codon <- gsub("T", "U", m_lucif_data$Codon)
a_lyrata_data$Codon <- gsub("T", "U", a_lyrata_data$Codon)
d_willistoni_data$Codon <- gsub("T", "U", d_willistoni_data$Codon)

olucim_box <-  boxoploto(o_lucim_data)
mlucif_box <-  boxoploto(m_lucif_data)
alyrata_box <-  boxoploto(a_lyrata_data)
dwilli_box <- boxoploto(d_willistoni_data)

olucim_box_zoom <-  boxoploto(o_lucim_data,0.9)
mlucif_box_zoom <-  boxoploto(m_lucif_data,0.9)
alyrata_box_zoom <-  boxoploto(a_lyrata_data,0.9)

olucim_dens <- densoploto(o_lucim_data)
mlucif_dens <- densoploto(m_lucif_data)
alyrata_dens <- densoploto(a_lyrata_data)
```

And some statistics

```{r}
o_lucim_data %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
m_lucif_data  %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
a_lyrata_data  %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
d_willistoni_data %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))

o_lucim_data %>% group_by(Codon) %>% dplyr::count()
m_lucif_data %>% group_by(Codon) %>% dplyr::count()
a_lyrata_data %>% group_by(Codon) %>% dplyr::count()
d_willistoni_data %>% group_by(Codon) %>% dplyr::count()
```
