---
title: "3'UTR length distribution among eukaryotes"
output: html_document
date: "2024"
---

```{r}
library(biomaRt)
library(AnnotationHub)
library(orthogene)
library(tidyverse)
library(ggridges)
library(ggpubr)
library(rstatix)
library(hrbrthemes)
library(dplyr)
library(readr)
library(DESeq2)
library(clusterProfiler)
library(GOfuncR)
library(GO.db)
```

Visualization functions

```{r}
boxoploto <- function(data_for_plot, zoom=1){
  data_for_plot %>% ggplot(aes(x = Codon, y = Length, color = Codon)) +
    geom_boxplot()+
    coord_cartesian(ylim = quantile(data_for_plot$Length, c(0, zoom)))+
    scale_color_manual(values = c("#CC3210","#F18905","#074258"))+
    theme_minimal()+
    theme(text = element_text(size = 15))
} #function for plotting alike boxplots with zoom option, enabling scaling. zoom takes values from 0 to 1 representing the part of the data to be plotted (from 0 to zoom)

densoploto <- function(data_for_plot,maxlength = 1000000000){
  data_for_plot %>% filter (Length < maxlength) %>% ggplot(aes(x=Length, color=Codon))+
  geom_density(adjust=1.5)+
  scale_color_manual(values = c("#CC3210","#F18905","#074258"))+
  theme_ipsum()
} #function for plotting alike probability density plots with maxlength option, enabling to set the limit of 3'UTR length to be included to the plot
```


##Human data analysis

###General distribution

```{r}
human_data <- read_delim("../data/Human.tsv",delim="\t", col_names = c("Transcript_ID", "Codon", "Length"))
human_data$Codon <- gsub("T", "U", human_data$Codon)

human_data%>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
human_data %>% group_by(Codon) %>% count()
```

```{r}
human_box <- boxoploto(human_data) 
human_box_09 <- boxoploto(human_data, 0.9)
human_dens <- densoploto(human_data)
human_dens_zoom <- densoploto(human_data, 2000)
```

###Human enrichment analysis

####Small 3'UTRs

Two groups: L<=10 and 10<L<=50

```{r}
hum_trans_gene <- merge(human_data, hum_total, by="Transcript_ID") %>% dplyr::select(Gene,Codon,Length)
hum_10 <- hum_trans_gene %>% filter(Length<=10)

hum_10_go <- enrichGO(gene = hum_10$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "ALL",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)

hum_10_go_tsv <- data.frame(hum_10_go)
write_tsv(hum_10_go_tsv, file = "../results/hum_10_go.tsv")

hum_11_50 <- hum_trans_gene %>% filter(Length>10 & Length <=50)
hum_11_50_go_CC <- enrichGO(gene = hum_11_50$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "CC",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)
hum_11_50_go_BP <- enrichGO(gene = hum_11_50$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "BP",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)
hum_11_50_go_MF <- enrichGO(gene = hum_11_50$Gene, 
                    universe=hum_trans_gene$Gene, 
                    keyType = "ENSEMBL",
                    ont = "MF",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)

hum_10_go_emap <- hum_10_go %>% enrichplot::pairwise_termsim() %>%
  emapplot(showCategory = 50,cex_label_category = 0.8, cex_category = 0.5, repel = TRUE)



hum_11_50_go_CC_emap <- hum_11_50_go_CC %>% enrichplot::pairwise_termsim() %>%
emapplot(showCategory = 50,cex_label_category = 0.5, cex_category = 0.5, repel = TRUE, label_format  = 0.5) 
hum_11_50_go_MF_emap <- hum_11_50_go_MF %>% enrichplot::pairwise_termsim() %>%
emapplot(showCategory = 50,cex_label_category = 0.5, cex_category = 0.5, repel = TRUE, label_format  = 0.5) 
hum_11_50_go_BP_emap <- hum_11_50_go_BP %>% enrichplot::pairwise_termsim() %>%
emapplot(showCategory = 50,cex_label_category = 0.45, cex_category = 0.5, repel = TRUE, label_format  = 0.5)

hum_11_50_go_CC_tsv <- data.frame(hum_11_50_go_CC)
write_tsv(hum_11_50_go_CC_tsv, file = "../results/hum_11_50_go_ะกะก.tsv")
hum_11_50_go_MF_tsv <- data.frame(hum_11_50_go_MF)
write_tsv(hum_11_50_go_MF_tsv, file = "../results/hum_11_50_go_MF.tsv")
hum_11_50_go_BP_tsv <- data.frame(hum_11_50_go_BP)
write_tsv(hum_11_50_go_BP_tsv, file = "../results/hum_11_50_go_BP.tsv")

```

And now let's obtain genes that are located in the nodes of interest (ribosomes, translation) and have only short 3'UTRs. Since single gene can have several transcript variants, it can have 3'UTRs <50 and >100. Here we are interested in the genes having only small variants. Those that have both short and long 3'UTRs are basically available from the results of enrichment analysis.  

First was to filter the letter files to obtain genes with only interesting GO categories and than get the list of unique genes, as they can occur in several GO categories. This was made with Python script ../scripts_analysis/gene_extractor.py

```{r}
hum_11_50_CC_genes <- read_delim ("../results/CC_GENES_50_HUM.tsv", delim = "\t") #there are also genes that also have long 3'UTRs, let's get rid of them
hum_11_50_MF_genes <- read_delim ("../results/MF_GENES_50_HUM.tsv", delim = "\t")
hum_11_50_BP_genes <- read_delim ("../results/BP_GENES_50_HUM.tsv", delim = "\t")

hum_11_50_CC_genes_ensembl <- getBM(c("external_gene_name","ensembl_gene_id"),"external_gene_name", hum_11_50_CC_genes$Gene, mart)
hum_11_50_MF_genes_ensembl <- getBM(c("external_gene_name","ensembl_gene_id"),"external_gene_name", hum_11_50_MF_genes$Gene, mart)
hum_11_50_BP_genes_ensembl <- getBM(c("external_gene_name","ensembl_gene_id"),"external_gene_name", hum_11_50_BP_genes$Gene, mart)

colnames(hum_11_50_CC_genes_ensembl) <-  c("Symbol", "Gene")
colnames(hum_11_50_BP_genes_ensembl) <-  c("Symbol", "Gene")
colnames(hum_11_50_MF_genes_ensembl) <-  c("Symbol", "Gene")

hum_11_50_CC_trans_ensembl <- merge(hum_11_50_CC_genes_ensembl, hum_trans_gene, by = "Gene")
hum_11_50_BP_trans_ensembl <- merge(hum_11_50_BP_genes_ensembl, hum_trans_gene, by = "Gene")
hum_11_50_MF_trans_ensembl <- merge(hum_11_50_MF_genes_ensembl, hum_trans_gene, by = "Gene") %>% merge(human_data, by = "Transcript_ID")

hum_11_50_CC_trans_length_ensembl <- merge(hum_11_50_CC_trans_ensembl, human_data, by = "Transcript_ID")
hum_11_50_BP_trans_length_ensembl <- merge(hum_11_50_BP_trans_ensembl, human_data, by = "Transcript_ID")
hum_11_50_MF_trans_length_ensembl <- merge(hum_11_50_MF_trans_ensembl, human_data, by = "Transcript_ID")

CC_less <-  hum_11_50_CC_trans_length_ensembl %>% filter (Length <= 50) 
CC_more <-  hum_11_50_CC_trans_length_ensembl %>% filter (Length > 50) 
CC_small <- CC_less %>% filter(!Symbol.x %in% CC_more$Symbol.x) %>% dplyr::select("Transcript_ID", "Symbol.x", "Length")

MF_less <-  hum_11_50_MF_trans_length_ensembl %>% filter (Length <= 50) 
MF_more <-  hum_11_50_MF_trans_length_ensembl %>% filter (Length > 50) 
MF_small <- MF_less %>% filter(!Symbol.x %in% MF_more$Symbol.x)%>% dplyr::select("Transcript_ID", "Symbol.x", "Length")

BP_less <-  hum_11_50_BP_trans_length_ensembl %>% filter (Length <= 50) 
BP_more <-  hum_11_50_BP_trans_length_ensembl %>% filter (Length > 50) 
BP_small <- BP_less %>% filter(!Symbol.x %in% BP_more$Symbol.x)%>% dplyr::select("Transcript_ID", "Symbol.x", "Length")

write_tsv(CC_small, "../results/CC_small.tsv") #there are genes with only short 3'UTRs
write_tsv(MF_small, "../results/MF_small.tsv")
write_tsv(BP_small, "../results/BP_small.tsv")
```

###Housekeeping genes

List of human housekeeping genes is taken from https://www.tau.ac.il/~elieis/HKG/. It is stored in the "../data/HK_genes.txt" file

```{r}
hum_house <- read_delim("../data/HK_genes.txt", delim = " \t", col_names = c("SYMBOL", "Nucl_DB"))

hum_house_ensembl <- getBM("ensembl_gene_id","external_gene_name", hum_house$SYMBOL, mart)
colnames(hum_house_ensembl) <- c("Gene")
hum_house_merged <- merge(hum_house_ensembl, hum_trans_gene, by="Gene")

hum_house_merged$Codon <- gsub("T", "U", hum_house_merged$Codon)

hum_house_box <- boxoploto(hum_house_merged)
hum_house_box_09 <- boxoploto(hum_house_merged, 0.9)
hum_house_dens <- densoploto(hum_house_merged)
hum_house_dens_zoom <- densoploto(hum_house_merged, 1500)

hum_house_merged %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
```


##Rabbit data analysis

###Rabbit vs Human


Primarily, it's required to obtain the list of rabbit and human orthologous genes. 

```{r}
mart <- useMart("ENSEMBL_MART_ENSEMBL","hsapiens_gene_ensembl") #human mart object
```

Get ENSEMBL gene IDs from transcript IDs
```{r}
rab_res <- read_delim("../data/Rabbit.tsv",delim="\t",col_names = c("Transcript_ID","Stop","Length"))
rab_total <- read_delim("../data/Oryctolagus.tsv",delim="\t",col_names = c("Symbol","Gene","Transcript_ID"))

merged_res <- merge(rab_res,rab_total,by="Transcript_ID")
rab_genes <- merged_res %>% select(Gene)
colnames(rab_genes) <- c("Rab_Gene")

```

```{r}
rab_mart <- useMart("ensembl", dataset = "ocuniculus_gene_ensembl", host="https://feb2021.archive.ensembl.org")
human_mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl",host = "https://feb2021.archive.ensembl.org")

orth <- getLDS(mart = rab_mart, attributes = c("ensembl_gene_id"), martL = human_mart, attributesL = c("ensembl_gene_id"))
colnames(orth) <- c("Rab_Gene", "Hum_Gene")

```

```{r}
target_orths <- merge(rab_genes, orth, by = "Rab_Gene") #have some replicated rows, so has to be filtered
```

Filtration was made using bash following writing target_orths to a tsv file. 

```{r}
uniq_orths <- read_delim("../data/uRH.tsv", delim= "\t")
target_genes_rabbit <- uniq_orths %>% dplyr::select(Rab_Gene)
target_genes_human <- uniq_orths %>% dplyr::select(Hum_Gene)
colnames(target_genes_rabbit) <- c("Gene")
colnames(target_genes_human) <- c("Gene")
```

Now we have a list of orthologous genes. Next step is to get the list of transcripts for each organism. To do that tables provided by the UTRdb 2.0 will be used

```{r}
hum_total <- read_delim("../data/Homo.tsv",delim="\t",col_names = c("Symbol", "Gene", "Transcript_ID")) #same table for rabbit has been already downloaded above to rab_total variable

target_rabbit_transcripts <- merge(target_genes_rabbit, rab_total, by = "Gene") %>% dplyr::select("Transcript_ID")
target_human_transcripts <- merge(target_genes_human, hum_total, by = "Gene") %>% dplyr::select("Transcript_ID") 

```

These variables (target_*_transcripts) were written to tsv files, filtered from replicated values final lists of transcripts were used to obtain information about 3'UTR lengths. This was made with Python script ../scripts_data/UTR_orth.py

Results are saved in the files RES_Human and RES_Rabbit. 
```{r}
hum_orth <- read_delim("../data/RES_Human.tsv",delim="\t",col_names = c("Trans_ID", "Codon","Length")) %>% dplyr::select (Codon,Length) %>% filter(Codon == "TGA" | Codon == "TAG" | Codon == "TAA")
rab_orth <- read_delim("../data/RES_Rabbit.tsv",delim="\t",col_names = c("Trans_ID", "Codon","Length"))%>% dplyr::select (Codon,Length)%>% filter(Codon == "TGA" | Codon == "TAG" | Codon == "TAA")

hum_orth$Codon <- gsub("T", "U",hum_orth$Codon )
rab_orth$Codon <- gsub("T", "U",rab_orth$Codon )

h_o_box <- boxoploto(hum_orth) 
r_o_box <- boxoploto(rab_orth)
h_o_box_zoomed <- boxoploto(hum_orth,0.9)
r_o_box_zoomed <- boxoploto(rab_orth,0.9)
h_o_dens <- densoploto(hum_orth,5000)
r_o_dens <- densoploto(rab_orth)

hum_orth %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
rab_orth %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))

hum_orth %>% group_by(Codon) %>% count()
rab_orth %>% group_by(Codon) %>% count()
```

Kruskal-Wallis test to define if stop-codon type affects the median 3'UTR length. 

```{r}
kruskal.test(Length ~ Codon, data = rab_orth)
kruskal.test(Length ~ Codon, data = hum_orth)
dunn_test(rab_orth, Length ~ Codon)

R_UAA_UAG <-rab_orth %>% filter (Codon == "TAG" | Codon == "TGA")
wilcox.test(Length ~ Codon, data = R_UAA_UAG, paired = FALSE)
```

###Rabbit enrichment analysis


###Second mode area

Since rabbit 3'UTR length distribution has second mode in the area of 1000 we'll check the transcripts with length 950<L<1050. The problem here is that rabbit database is poorly annotated. To solve it we'll use the annotation of orthologous human genes, which are annotated much more better

```{r}
rab_1000 <- rab_orth%>% filter(Codon == "TGA" | Codon == "TAG" | Codon == "TAA") %>% filter (Length < 1050 & Length > 950)
colnames(rab_1000) <- c("Transcript_ID", "Codon","Length")

merged_1000 <- merge(rab_1000, merged_res, by = "Transcript_ID") %>% dplyr::select(Length.x, Gene)
colnames(merged_1000) <-c("Length", "Rab_Gene")
hum_1000 <- merge(merged_1000, uniq_orths, by = "Rab_Gene") %>% unique() %>% dplyr::select(Hum_Gene)
hum_bck <- uniq_orths %>% dplyr::select(Hum_Gene) %>% unique() #this will be used as a universe  parameter value, background. 
```

Now, obtaining their GO categories through enrichment analysis

```{r}
go_1000 <- enrichGO(gene = hum_1000$Hum_Gene, 
                    universe=hum_bck$Hum_Gene, 
                    keyType = "ENSEMBL",
                    ont = "ALL",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)
RAB_GO1000 <- data.frame(go_1000)
write_tsv(RAB_GO1000, file = "../results/RAB_GO1000.tsv")

```

And let's visualize them

```{r}
go_1000 %>%  
  enrichplot::pairwise_termsim() %>%
  emapplot(showCategory = 50)->enrichplot_1000
```


####Small 3'UTRs
The same procedure for small 3'UTRs, less than 50 nt

```{r}
rab_50 <- rab_orth %>% filter(Codon == "TGA" | Codon == "TAG" | Codon == "TAA") %>% filter (Length < 50)
colnames(rab_50) <- c("Transcript_ID", "Codon","Length")

merged_50 <- merge(rab_50, merged_res, by = "Transcript_ID") %>% dplyr::select(Length.x, Gene)
colnames(merged_50) <-c("Length", "Rab_Gene")
hum_50 <- merge(merged_50, uniq_orths, by = "Rab_Gene") %>% unique() %>% dplyr::select(Hum_Gene)

```

```{r}
go_50 <- enrichGO(gene = hum_50$Hum_Gene, 
                    universe=hum_bck$Hum_Gene, 
                    keyType = "ENSEMBL",
                    ont = "ALL",
                    OrgDb ="org.Hs.eg.db",
                    pAdjustMethod = "BH", 
                    readable = TRUE)

RAB_GO50 <- data.frame(go_50)
write_tsv(RAB_GO50, file = "../results/RAB_GO50.tsv")

go_50 %>%  
  enrichplot::pairwise_termsim() %>%
  emapplot(showCategory = 50)->enrichplot_50
```

##General distribution

```{r}
total_prec <- read_delim(unz("../data/TOTAL_PY.tsv.zip","TOTAL_PY.tsv"),delim = "\t") #TOTAL_PY.tsv is too big, so it's kept as a .zip file in the repo
colnames(total_prec) <- c("UAA", "UAG", "UGA")
prec_plot <-  total_prec %>% pivot_longer(cols = codons, names_to = "Codon", values_to = "Length") %>% drop_na()
prec_dens <- densoploto(prec_plot)
prec_box <- boxoploto(prec_plot)
prec_box_zoom <- boxoploto(prec_plot, 0.9)
prec_dens_zoom <- densoploto(prec_plot, 2000)

prec_plot %>% nrow()
prec_plot %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
prec_plot %>% group_by(Codon) %>% dplyr::count()
```


Statistical evaluation
```{r}
kruskal.test(Length ~ Codon, data = prec_plot)
dunn_test(prec_plot, Length ~ Codon)
```



##Small 3'UTR holders

###Paramecium tetraurelia

Data is obtained with Python script ../scripts_data/Paramecium.py

```{r}
param_data <- read_delim("../data/Paramecium_tetraurelia.RES.tsv", delim="\t", col_names = c("Transcript_ID", "Codon", "Length"))

param_data$Codon <- gsub("T", "U", param_data$Codon)

param_box <- boxoploto(param_data %>% filter (Codon =="UGA"))
param_box_nonzero <- boxoploto(param_data %>% filter (Length> 0))
param_box_nonzero_95 <- boxoploto(param_data %>% filter (Length> 0), 0.95)
param_dens <- densoploto(param_data %>% filter (Codon =="UGA"))

param_data %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
param_data %>% group_by(Codon) %>% dplyr::count()

param_tga <- param_data %>% filter(Codon == "UGA")
```

###Others
Organisms having small 3'UTR length median values were chosen with Python script ../scripts_analysis/small_utrs.py

```{r}

codons_T <- c("TAA", "TAG", "TGA")
o_lucim_data <- read_delim("../data/Ostreococcus_lucimarinus.ASM9206v1.54.utrs.tsv", delim = "\t") %>% pivot_longer(cols = codons_T, names_to = "Codon", values_to = "Length") %>% drop_na()
m_lucif_data <- read_delim("../data/Myotis_lucifugus.Myoluc2.0.107.utrs.tsv", delim = "\t") %>% pivot_longer(cols = codons_T, names_to = "Codon", values_to = "Length") %>% drop_na()
a_lyrata_data <- read_delim("../data/Arabidopsis_lyrata.v.1.0.54.utrs.tsv", delim = "\t") %>% pivot_longer(cols = codons_T, names_to = "Codon", values_to = "Length") %>% drop_na()
o_lucim_data$Codon <- gsub("T", "U", o_lucim_data$Codon)
m_lucif_data$Codon <- gsub("T", "U", m_lucif_data$Codon)
a_lyrata_data$Codon <- gsub("T", "U", a_lyrata_data$Codon)

olucim_box <-  boxoploto(o_lucim_data)
mlucif_box <-  boxoploto(m_lucif_data)
alyrata_box <-  boxoploto(a_lyrata_data)

olucim_box_zoom <-  boxoploto(o_lucim_data,0.9)
mlucif_box_zoom <-  boxoploto(m_lucif_data,0.9)
alyrata_box_zoom <-  boxoploto(a_lyrata_data,0.9)

olucim_dens <- densoploto(o_lucim_data)
mlucif_dens <- densoploto(m_lucif_data)
alyrata_dens <- densoploto(a_lyrata_data)
```

And some statistics

```{r}
o_lucim_data %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
m_lucif_data  %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))
a_lyrata_data  %>% group_by(Codon) %>% summarise(Mean = mean(Length), Max = max(Length), Min = min(Length), Median = median(Length), Std = sd(Length))


o_lucim_data %>% group_by(Codon) %>% dplyr::count()
m_lucif_data %>% group_by(Codon) %>% dplyr::count()
a_lyrata_data %>% group_by(Codon) %>% dplyr::count()
```

